com.microsoft.java.debug.plugin/src/main/java/com/microsoft/java/debug/plugin/internal/JdtUtils.java

```diff
@@ -78,4 +98,164 @@ public static IJavaProject getJavaProject(String projectName) {
         IProject project = root.getProject(projectName);
         return getJavaProject(project);
     }
+
+    /**
+     * Given the project name, return the corresponding project object.
+     * If the project doesn't exist, return null.
+     */
+    public static IProject getProject(String projectName) {
+        if (projectName == null) {
+            return null;
+        }
+        IWorkspaceRoot root = ResourcesPlugin.getWorkspace().getRoot();
+        return root.getProject(projectName);
+    }
+
+    /**
+     * Compute the possible source containers that the current project could be associated with.
+     * <p>
+     * If current project is specified, try to parse the source containers from current project's classpath entries first,
+     * then the other projects at the same workspace.
+     * </p>
+     * <p>
+     * Otherwise, loop every projects at the current workspace and combine the parsed source containers directly.
+     * </p>
+     * @param currentProject
+     *                  the current project name.
+     * @return the possible source container list.
+     */
+    public static ISourceContainer[] getSourceContainers(String currentProject) {
+        Set<ISourceContainer> containers = new LinkedHashSet<>();
+        List<IProject> projects = new ArrayList<>();
+
+        IProject targetProject = JdtUtils.getProject(currentProject);
+        if (targetProject != null) {
+            projects.add(targetProject);
+        }
+
+        List<IProject> workspaceProjects = Arrays.asList(ResourcesPlugin.getWorkspace().getRoot().getProjects());
+        projects.addAll(workspaceProjects.stream().filter((project) -> {
+            return !project.equals(targetProject);
```

Comment:
Maybe we can just use whatever returned by getProjects() instead of filtering out the targetProject. The logic seems redundant. The project collection should already have the target project. So this function can have no parameter.

Reply:
The original purpose is `"If the project name is specified, it will put the source containers from the specified project in the front of the result. "`

The filter logic is to remove duplicated project to avoid duplicated computation.

Yes, it's unnecessary, because the final source container result is a set object, it'll remove the duplicated source container. The following code is OK.
<pre>
        IProject targetProject = JdtUtils.getProject(projectName);
        if (targetProject != null) {
            projects.add(targetProject);
        }

        List<IProject> workspaceProjects = Arrays.asList(ResourcesPlugin.getWorkspace().getRoot().getProjects());
        projects.addAll(workspaceProjects);
</pre>